Glooctl install federation demo

ashleywang1 opened the pull request.

joekelley commented on docs/content/reference/cli/glooctl_install_federation_demo.md:
````diff
@@ -0,0 +1,49 @@
+  -h, --help   help for demo
+```
+
+### Options inherited from parent commands
````
many of the inherited flags from install are not relevant here (e.g. dry-run, create namespace)

since this command will bootstrap kind clusters for the user, rather than install gloo-fed on existing clusters, what do you think about breaking it out into a different part of the command tree? Something like `demo federation`. I could see the `demo` part of the tree growing to include other scenarios we might want to feature to users.

ashleywang1 replied:
Yup, agreed.
`glooctl demo federation` makes a lot of sense.

ashleywang1 commented on projects/gloo/cli/pkg/cmd/install/federation.go:
```diff
@@ -39,7 +43,569 @@ func glooFedCmd(opts *options.Options) *cobra.Command {
+
+glooctl create secret tls --name failover-downstream --certchain tls.crt --privatekey tls.key --rootca mtls.crt
+
+# Apply failover gateway and service
```
Added a note

ashleywang1 commented:
/skip-changelog

ashleywang1 commented:
Here's the output for running the command:
```
✗ glooctl demo federation
+ '[' local == '' ']'
+ '[' remote == '' ']'
+ controlPlaneVersion=0.0.9
+ '[' 0.0.9 == '' ']'
+ printf 'control plane components will be installed with version %s\n' 0.0.9
control plane components will be installed with version 0.0.9
+ kind create cluster --name local
Creating cluster "local" ...
 ✓ Ensuring node image (kindest/node:v1.17.0) 🖼
 ✓ Preparing nodes 📦
 ✓ Writing configuration 📜
 ✓ Starting control-plane 🕹️
 ✓ Installing CNI 🔌
 ✓ Installing StorageClass 💾
Set kubectl context to "kind-local"
You can now use your cluster with:

kubectl cluster-info --context kind-local

Have a nice day! 👋
+ cat
+ kind create cluster --name remote --config=-
Creating cluster "remote" ...
 ✓ Ensuring node image (kindest/node:v1.17.0) 🖼
 ✓ Preparing nodes 📦
 ✓ Writing configuration 📜
 ✓ Starting control-plane 🕹️
 ✓ Installing CNI 🔌
 ✓ Installing StorageClass 💾
Set kubectl context to "kind-remote"
You can now use your cluster with:

kubectl cluster-info --context kind-remote

Have a nice day! 👋
+ kubectl config use-context kind-local
Switched to context "kind-local".
+ glooctl install federation
Creating namespace gloo-fed... Done.
Starting Gloo Federation installation...

Gloo Federation was successfully installed!
+ kubectl -n gloo-fed rollout status deployment gloo-fed --timeout=1m
Waiting for deployment "gloo-fed" rollout to finish: 0 of 1 updated replicas are available...
deployment "gloo-fed" successfully rolled out
+ kubectl config use-context kind-remote
Switched to context "kind-remote".
+ cat
+ glooctl install gateway --values nodeport.yaml
Creating namespace gloo-system... Done.
Starting Gloo installation...

Gloo was successfully installed!
+ rm nodeport.yaml
+ kubectl -n gloo-system rollout status deployment gloo --timeout=2m
Waiting for deployment "gloo" rollout to finish: 0 of 1 updated replicas are available...
deployment "gloo" successfully rolled out
+ kubectl -n gloo-system rollout status deployment discovery --timeout=2m
Waiting for deployment "discovery" rollout to finish: 0 of 1 updated replicas are available...
deployment "discovery" successfully rolled out
+ kubectl -n gloo-system rollout status deployment gateway-proxy --timeout=2m
deployment "gateway-proxy" successfully rolled out
+ kubectl -n gloo-system rollout status deployment gateway --timeout=2m
deployment "gateway" successfully rolled out
+ kubectl patch settings -n gloo-system default --type=merge -p '{"spec":{"watchNamespaces":["gloo-system", "default"]}}'
settings.gloo.solo.io/default patched
+ openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout tls.key -out tls.crt -subj /CN=solo.io
Generating a 2048 bit RSA private key
.............+++
.+++
writing new private key to 'tls.key'
-----
+ openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout mtls.key -out mtls.crt -subj /CN=solo.io
Generating a 2048 bit RSA private key
.................................+++
......+++
writing new private key to 'mtls.key'
-----
+ glooctl create secret tls --name failover-downstream --certchain tls.crt --privatekey tls.key --rootca mtls.crt
+---------------------+------+
|       SECRET        | TYPE |
+---------------------+------+
| failover-downstream | TLS  |
+---------------------+------+
+ kubectl apply -f -
gateway.gateway.solo.io/failover-gateway created
service/failover created
+ kubectl config use-context kind-local
Switched to context "kind-local".
+ cat
+ glooctl install gateway enterprise --values basic-enterprise.yaml
Creating namespace gloo-system... Done.
Starting Gloo Enterprise installation...

Gloo Enterprise was successfully installed!
+ rm basic-enterprise.yaml
+ kubectl -n gloo-system rollout status deployment gloo --timeout=2m
Waiting for deployment "gloo" rollout to finish: 0 of 1 updated replicas are available...
deployment "gloo" successfully rolled out
+ kubectl -n gloo-system rollout status deployment discovery --timeout=2m
deployment "discovery" successfully rolled out
+ kubectl -n gloo-system rollout status deployment gateway-proxy --timeout=2m
Waiting for deployment "gateway-proxy" rollout to finish: 0 of 1 updated replicas are available...
deployment "gateway-proxy" successfully rolled out
+ kubectl -n gloo-system rollout status deployment gateway --timeout=2m
deployment "gateway" successfully rolled out
+ glooctl create secret tls --name failover-upstream --certchain mtls.crt --privatekey mtls.key
----------
glooctl binary version (1.5.0-beta5) differs from server components (v1.4.5) by at least a minor version.
Consider running:
glooctl upgrade --release=v1.4.5
----------

----------
glooctl binary version (1.5.0-beta5) differs from server components (v1.4.5) by at least a minor version.
Consider running:
glooctl upgrade --release=v1.4.5
----------

+-------------------+------+
|      SECRET       | TYPE |
+-------------------+------+
| failover-upstream | TLS  |
+-------------------+------+
+ rm mtls.key mtls.crt tls.crt tls.key
+ case $(uname) in
++ uname
+ CLUSTER_DOMAIN_MGMT=host.docker.internal
+ CLUSTER_DOMAIN_REMOTE=host.docker.internal
+ glooctl cluster register --cluster-name kind-local --remote-context kind-local --local-cluster-domain-override host.docker.internal --federation-namespace gloo-fed
----------
glooctl binary version (1.5.0-beta5) differs from server components (v1.4.5) by at least a minor version.
Consider running:
glooctl upgrade --release=v1.4.5
----------

+ glooctl cluster register --cluster-name kind-remote --remote-context kind-remote --local-cluster-domain-override host.docker.internal --federation-namespace gloo-fed
----------
glooctl binary version (1.5.0-beta5) differs from server components (v1.4.5) by at least a minor version.
Consider running:
glooctl upgrade --release=v1.4.5
----------

+ echo 'Registered gloo clusters kind-local and kind-remote'
Registered gloo clusters kind-local and kind-remote
+ echo 'Set up resources for Failover demo...'
Set up resources for Failover demo...
+ kubectl apply -f -
service/service-blue created
deployment.apps/echo-blue-deployment created
+ kubectl config use-context kind-remote
Switched to context "kind-remote".
+ kubectl apply -f -
service/service-green created
deployment.apps/echo-green-deployment created
+ kubectl config use-context kind-local
Switched to context "kind-local".
+ kubectl apply -f -
federatedupstream.fed.gloo.solo.io/default-service-blue created
federatedvirtualservice.fed.gateway.solo.io/simple-route created
failoverscheme.fed.solo.io/failover-test-scheme created
+ cat

# We now have failover set up correctly!

# To view the federated upstreams, run:
kubectl get federatedupstream -n gloo-fed

# To view the federated virtual services, run:
kubectl get federatedvirtualservices -n gloo-fed

# To view the failover schemes, run:
kubectl get failoverschemes -n gloo-fed

# For this section, use two terminals, one for the port-forward command and one for the curl command.

# Curl the route to reach the blue pod. You should see a return value of "blue-pod".
kubectl port-forward -n gloo-system svc/gateway-proxy 8080:80
curl localhost:8080/

# Force the health check to fail
k port-forward deploy/echo-blue-deployment 19000
curl -X POST  localhost:19000/healthcheck/fail

# See that the green pod is now being reached, with the curl command returning "green-pod".
kubectl port-forward -n gloo-system svc/gateway-proxy 8080:80
curl localhost:8080/
+ cat

# To clean up the demo, run:
kind delete cluster --name "local"
kind delete cluster --name "remote"
```
